### ISSUE TITLE
Websocket response fails and halts the entire node.

### ISSUE THREAD

**Issue by intl-man** [tendermint/tendermint#6184]
**Tendermint version**:

v0.33.9

**What happened**:

When a websocket client is no longer active, but the node attempts to send a response over the wire anyway, we witness the following and the node then stops processing blocks. No panic nor any other errors to indicate that something has gone horribly wrong.

```
I[2021-02-25|23:41:18.352] Committed state                              module=state height=224327 txs=0 appHash=24667534E5145D0B6707362E21AA4F2629E41DC25D7F569A3C8D318392E513E5
I[2021-02-25|23:41:24.043] Executed block                               module=state height=224328 validTxs=0 invalidTxs=0
I[2021-02-25|23:41:24.053] Committed state                              module=state height=224328 txs=0 appHash=F87F93A004E5640D4928B0C4319483B35966820A8C6D06C39F0C7B9776C42322
I[2021-02-25|23:41:29.587] Executed block                               module=state height=224329 validTxs=0 invalidTxs=0
I[2021-02-25|23:41:29.596] Committed state                              module=state height=224329 txs=0 appHash=2E04F4DE63347C1A1DF29EBB2F4279CD2FB0703734A5A9D27E0CD31DC537BE78
E[2021-02-25|23:41:29.679] Failed to read request                       module=rpc-server protocol=websocket remote=10.0.50.12:60919 err="websocket: close 1001 (going away)"
E[2021-02-25|23:41:29.695] Failed to write response                     module=rpc-server protocol=websocket remote=10.0.50.12:60919 msg="RPCResponse{1 [123 34 113 117 101 114 121 34 58 34 116 109 46 101 118 101 110 116 61 39 78 101 119 66 108 111 99 107 39 34 44 34 100 97 116 97 34 58 123 34 116 121 112 101 34 58 34 116 101 110 100 101 114 109 ....
```

**What you expected to happen**:

That it wouldn't attempt to send over a closed connection and if it did, it wouldn't hang/lock up.

**Have you tried the latest version**: yes/no

No.

**How to reproduce it** (as minimally and precisely as possible):

Make hundreds of websocket requests over a few hours.

**Comment by melekes** [comment]
Could you send us the goroutine list of a hanged node? You can get it by killing it with `kill -6 <PID>`, where <PID> is the process ID of Tendermint (`ps aux | grep tendermint`)

**Comment by melekes** [comment]
Possible duplicate of https://github.com/tendermint/tendermint/issues/6147

**Comment by litvintech** [comment]
Me with @mrlp4 confirming that issue. In our euler-6 network + cyber.page app we got it ~6 times during a year.
That's maybe related to peak usage of app

**Comment by melekes** [comment]
Can't replicate this w/

```go
package main

import (
	"context"
	"flag"
	"log"
	"time"

	"github.com/tendermint/tendermint/rpc/client/http"
)

const subscriber = "disconnecting-client"

var addr string
var nclients int

func init() {
	flag.StringVar(&addr, "addr", "", "address to connect")
	flag.IntVar(&nclients, "nclients", 1000, "number of clients")
}

func main() {
	flag.Parse()

	for i := 0; i < nclients; i++ {
		c, err := http.New(addr)
		if err != nil {
			log.Fatal(err)
		}
		err = c.Start()
		if err != nil {
			log.Fatal(err)
		}

		out, err := c.Subscribe(context.Background(), subscriber, "tm.event='NewBlock'")
		if err != nil {
			log.Fatal(err)
		}

		go func(client int) {
			for {
				select {
				case e := <-out:
					log.Printf("%d client got %v", client, e)
				}
			}
		}(i)

	}

	time.Sleep(30 * time.Second)
}
```

I've tried to run this 20 times and interrupting w/ CTRL-C, but no results.

**Comment by Yurist-85** [comment]
https://github.com/tendermint/tendermint/blob/8175b2b26dee0117d617568bf12350c4b3698a1f/rpc/jsonrpc/server/ws_handler.go#L403-L413

Why don't we `continue` on line 411 instead of `return`?
It should solve the problem with hanging.

**Comment by creachadair** [comment]
> Why don't we `continue` on line 411 instead of `return`? It should solve the problem with hanging.

Returning is what we want here: Returning from the write thread unblocks the `Start` call on the connection, and returns control to the HTTP handler, which will then shut it down.

(It's possible the plumbing was different in previous versions, but that is what should happen in v0.35.x and forward)
